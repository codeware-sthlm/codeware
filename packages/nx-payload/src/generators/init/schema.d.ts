import type { Schema } from '@nx/node/src/generators/init/schema';

export type InitSchema = Schema;
