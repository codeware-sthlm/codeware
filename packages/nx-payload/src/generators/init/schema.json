{
  "title": "Init Nx Payload Plugin",
  "type": "object",
  "properties": {},
  "required": []
}
