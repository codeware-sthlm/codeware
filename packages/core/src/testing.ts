export { createSchemaTests } from './lib/testing/create-schema-tests';
export { SchemaRegistry } from './lib/testing/schema-registry.class';
