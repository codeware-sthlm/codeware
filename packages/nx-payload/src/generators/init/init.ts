import {
  type GeneratorCallback,
  type Tree,
  createProjectGraphAsync,
  formatFiles,
  readNxJson,
  runTasksInSerial
} from '@nx/devkit';
import { addPlugin } from '@nx/devkit/src/utils/add-plugin';

import type { PayloadTarget } from '../../utils/definitions';
import { isPluginInferenceEnabled } from '../../utils/is-plugin-inference-enabled';

import { updateDependencies } from './libs/update-dependencies';
import type { InitSchema } from './schema';

/**
 * Install Payload dependencies.
 */
export async function initGenerator(
  host: Tree,
  schema: InitSchema
): Promise<GeneratorCallback> {
  return initGeneratorInternal(host, { addPlugin: false, ...schema });
}

/**
 * @internal
 * Defined as init factory function in `generators.json`
 */
export async function initGeneratorInternal(
  host: Tree,
  schema: InitSchema
): Promise<GeneratorCallback> {
  const tasks: Array<GeneratorCallback> = [];

  const nxJson = readNxJson(host);
  // Only use inference state if not explicitly set
  schema.addPlugin ??= isPluginInferenceEnabled(nxJson);

  // Install NextJS dependencies and add plugin when inference is enabled

  // Install Payload dependencies
  const installTask = updateDependencies(host);
  tasks.push(installTask);

  if (schema.addPlugin) {
    const { createNodesV2 } = await import('../../plugins/plugin');

    await addPlugin(
      host,
      await createProjectGraphAsync(),
      '@cdwr/nx-payload/plugin',
      createNodesV2,
      {
        generateTargetName: ['gen' satisfies PayloadTarget],
        payloadTargetName: ['payload' satisfies PayloadTarget],
        payloadGraphqlTargetName: ['payload-graphql' satisfies PayloadTarget],
        dxMongodbTargetName: ['dx:mongodb' satisfies PayloadTarget],
        dxPostgresTargetName: ['dx:postgres' satisfies PayloadTarget],
        dxStartTargetName: ['dx:start' satisfies PayloadTarget],
        dxStopTargetName: ['dx:stop' satisfies PayloadTarget]
      },
      false
    );
  }

  if (!schema.skipFormat) {
    await formatFiles(host);
  }

  return runTasksInSerial(...tasks);
}
