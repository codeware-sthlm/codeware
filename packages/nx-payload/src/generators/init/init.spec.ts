import { type Tree, addDependenciesToPackageJson, readJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import type { PackageJson } from 'nx/src/utils/package-json';

import { initGenerator } from './init';

describe('init', () => {
  let tree: Tree;

  console.log = jest.fn();
  console.warn = jest.fn();

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should generate correct dependencies in package.json', async () => {
    await initGenerator(tree, {});
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    // Versions number can be updated by pipelines or similar tools,
    // so we only match for semantic patterns.
    const dynamicVersion = /^[^|~]?\d+\.\d+\.\d+$/;
    expect(packageJson).toMatchSnapshot({
      dependencies: {
        '@payloadcms/bundler-webpack': expect.stringMatching(dynamicVersion),
        '@payloadcms/db-mongodb': expect.stringMatching(dynamicVersion),
        '@payloadcms/db-postgres': expect.stringMatching(dynamicVersion),
        '@payloadcms/richtext-slate': expect.stringMatching(dynamicVersion),
        payload: expect.stringMatching(dynamicVersion),
        zod: 'latest'
      },
      devDependencies: {
        '@nx/node': expect.stringMatching(/^\d+\.\d+\.\d+$/)
      }
    });
  });

  it('should keep existing dependencies', async () => {
    const existing = 'existing';
    const existingVersion = '1.0.0';
    const dependencies = {
      [existing]: existingVersion
    };
    const devDependencies = {
      [existing]: existingVersion
    };
    addDependenciesToPackageJson(tree, dependencies, devDependencies);

    await initGenerator(tree, {});
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies[existing]).toBeDefined();
    expect(packageJson.devDependencies[existing]).toBeDefined();
  });
});
