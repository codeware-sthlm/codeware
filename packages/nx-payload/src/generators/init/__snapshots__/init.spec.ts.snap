// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init should generate correct dependencies in package.json 1`] = `
{
  "dependencies": {
    "@payloadcms/bundler-webpack": StringMatching /\\^\\[\\^\\|~\\]\\?\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
    "@payloadcms/db-mongodb": StringMatching /\\^\\[\\^\\|~\\]\\?\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
    "@payloadcms/db-postgres": StringMatching /\\^\\[\\^\\|~\\]\\?\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
    "@payloadcms/richtext-slate": StringMatching /\\^\\[\\^\\|~\\]\\?\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
    "express": StringMatching /\\^\\[\\^\\|~\\]\\?\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
    "payload": StringMatching /\\^\\[\\^\\|~\\]\\?\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
  },
  "devDependencies": {
    "@nx/express": StringMatching /\\^\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\$/,
  },
  "name": "@proj/source",
}
`;

